### PR TITLE
dialects: Add FastMathFlagsAttr to the Arith dialect.

### PR DESCRIPTION
**Pull request by KarelPeeters**
I needed support for the [`arith.fastmath` attribute](https://mlir.llvm.org/docs/Dialects/ArithOps/#fastmathflagsattr), so I added it based on the code [here](https://github.com/llvm/llvm-project/blob/ee9a0f30ca8a3582d0738f7499ea902e1f713b39/mlir/include/mlir/Dialect/Arith/IR/ArithBase.td#L96-L127).

* Are PR adding small details like this to the dialects welcome?
* This is one of my first dialect additions, did I do it right?

### CODE REVIEW

**Comment by superlopuh** [comment]
Thank you for adding these! Could you please add some tests for this? Potentially in the filecheck suite.

**Comment by KarelPeeters** [comment]
Sure, I can try writing some filecheck tests. `lit tests/filecheck` doesn't seem for work for me though, 59/69 tests silently fail. Is there a way to get it to sell my why they're failing?

Should I also add manual parser tests somewhere? I've been doing stuff like this, is there a good place to put those?

```python
source = """
"builtin.module"() ({
    %0 = "arith.constant"() {value = 1.000000e+00 : f32} : () -> f32
    %1 = "arith.constant"() {value = 2.000000e+00 : f32} : () -> f32
    %2 = "arith.mulf"(%0, %1) {fastmath = #arith.fastmath<nsz>} : (f32, f32) -> f32
}) : () -> ()
"""

ctx = build_ctx()
parser = MLIRParser(ctx=ctx, input=source, name="<unknown>", allow_unregistered_ops=True)
module = parser.parse_module()

flags = cast(FastMathFlagsAttr, module.ops[2].attributes["fastmath"])
assert flags.data == FastMathFlags({FastMathFlag.NO_SIGNED_ZEROS})
```

**Comment by superlopuh** [comment]
That doesn't seem quite right, I'd recommend constructing the operation directly, checking that it was instantiated correctly, printing it using `Printer(target=Printer.Target.MLIR)`, and parsing it to see that they are structurally equivalent.

**Comment by KarelPeeters** [comment]
> That doesn't seem quite right, I'd recommend constructing the operation directly, checking that it was instantiated correctly, printing it using `Printer(target=Printer.Target.MLIR)`, and parsing it to see that they are structurally equivalent.

If you do it this way you're never actually testing that the MLIR parse/print format is correct though, just that they match.

**Review by math-fehr — CHANGES_REQUESTED** [review]
Nice! The attribute looks good to me, we just need tests for this!

The way I would do this is just adding an `.xdsl` and `.mlir` test in the repo, at location [xdsl](https://github.com/xdslproject/xdsl)/[tests](https://github.com/xdslproject/xdsl/tree/main/tests)/[filecheck](https://github.com/xdslproject/xdsl/tree/main/tests/filecheck)/[arith_ops.xdsl](https://github.com/xdslproject/xdsl/blob/main/tests/filecheck/arith_ops.xdsl), and [tests](https://github.com/xdslproject/xdsl/tree/main/tests)/[filecheck](https://github.com/xdslproject/xdsl/tree/main/tests/filecheck)/[mlir-conversion](https://github.com/xdslproject/xdsl/tree/main/tests/filecheck/mlir-conversion)/[with-mlir](https://github.com/xdslproject/xdsl/tree/main/tests/filecheck/mlir-conversion/with-mlir)/[dialects](https://github.com/xdslproject/xdsl/tree/main/tests/filecheck/mlir-conversion/with-mlir/dialects)/arith/arith_attrs.mlir .

Can you show us which error do you get with lit ? You can add `-v` to get the errors.

**Comment by superlopuh** [comment]
> > That doesn't seem quite right, I'd recommend constructing the operation directly, checking that it was instantiated correctly, printing it using `Printer(target=Printer.Target.MLIR)`, and parsing it to see that they are structurally equivalent.
> 
> If you do it this way you're never actually testing that the MLIR parse/print format is correct though, just that they match.

Yep, the test that it's right is with filecheck instead.

**Comment by webmiche** [comment]
Thanks a lot for your PR, these changes are very welcome!

I am not sure how extensively these kind of enum-style attritues are used in MLIR, but I feel it might make sense to add some framework support for them. Maybe @math-fehr has an idea?

Especially the FastMathFlags class in this PR could be generalized to something that holds a set of any Enum.

I am also not quite sure how they are handled in MLIR, so maybe an ArrayAttr could have done it as well. Would be great to see some tests, for line coverage as well as for understanding of the MLIR representation.

**Comment by math-fehr** [comment]
> Thanks a lot for your PR, these changes are very welcome!
> 
> I am not sure how extensively these kind of enum-style attritues are used in MLIR, but I feel it might make sense to add some framework support for them. Maybe @math-fehr has an idea?
> 
> Especially the FastMathFlags class in this PR could be generalized to something that holds a set of any Enum.
> 
> I am also not quite sure how they are handled in MLIR, so maybe an ArrayAttr could have done it as well. Would be great to see some tests, for line coverage as well as for understanding of the MLIR representation.

Yes, I agree we should start supporting enums (as a subclass of Data I guess). Though for this PR it's not needed I would feel

**Comment by KarelPeeters** [comment]
Thanks for all the feedback!

The problem with `lit` was that the [`xdsl-opt` script](https://github.com/xdslproject/xdsl/blob/main/xdsl/tools/xdsl-opt) doesn't work on windows. I was using a workaround to run it myself but that didn't help `lit`.

I've now added some filecheck tests. They revealed the output order of attributes was non-consistent between runs (because of the `set`), which I have fixed.

It looks like the xDSL language now accepts both `!arith.fastmath<nsz>` and `#arith.fastmath<nsz>`, is that normal? `FastMathFlagsAttr` doesn't declare itself as a type as far as I understand.

**Comment by webmiche** [comment]
Sidecomment: if you have a good workaround for windows, I would very much appreciate a PR. Getting better support for other OS is crucial IMO.

**Comment by KarelPeeters** [comment]
> Sidecomment: if you have a good workaround for windows, I would very much appreciate a PR. Getting better support for other OS is crucial IMO.

I looked into this and it was an easy fix, see #475.

**Review by superlopuh — CHANGES_REQUESTED** [review]
Changes mostly look good, could you please address the pyright errors? I'd recommend setting up pyright locally to see them in vscode or your preferred IDE

**Comment by KarelPeeters** [comment]
I think I've fixed them, running  `pyright -p .\pyright-ci.json` locally doesn't show any errors any more.

**Comment by KarelPeeters** [comment]
I think I've fixed them, running  `pyright -p .\pyright-ci.json` locally doesn't show any errors any more.

**Review by math-fehr — APPROVED** [review]
Thanks! That looks good to me!

**Comment by math-fehr** [comment]
@superlopuh, can you check if this is good for you? So we can merge it

File: xdsl/dialects/arith.py [hunk 1]
```diff
@@ -650,4 +716,6 @@ def get(cls, input: SSAValue | Operation, target_type: Attribute):
 
         # Casts
         IndexCastOp,
-], [])
+], [
+        FastMathFlagsAttr
```
Comment (superlopuh) [inline xdsl/dialects/arith.py:720]: nit:
```suggestion
        FastMathFlagsAttr,
```

File: xdsl/dialects/arith.py [hunk 2]
```diff
@@ -605,6 +611,66 @@ def get(cls, input: SSAValue | Operation, target_type: Attribute):
...
+        return result
+
+    @staticmethod
+    def print_parameter(data: FastMathFlags, printer: Printer):
+        if len(data.flags) == 0:
+            printer.print("none")
+        elif len(data.flags) == len(FastMathFlag):
+            printer.print("fast")
+        else:
+            # make sure we emits flags in a consistent order
```
Comment (superlopuh) [inline xdsl/dialects/arith.py:666]: nit:
```suggestion
            # make sure we emit flags in a consistent order
```

**Review by superlopuh — APPROVED** [review]

**Pull request merged**